domain/consensus/processes/pruningmanager/pruningmanager.go

```diff
@@ -1,51 +1,226 @@
+		}
+		pm.pruningStore.Stage(candidatePHash, serializedUtxo)
+		currentP = candidatePHash
+	}
+	return pm.deletePastBlocks(currentP)
 }
 
 // PruningPoint returns the hash of the current pruning point
 func (pm *pruningManager) PruningPoint() (*externalapi.DomainHash, error) {
-	return nil, nil
+	pruningPoint, err := pm.pruningStore.PruningPoint(pm.databaseContext)
+	if err != nil {
+		if !pm.pruningStore.HasPruningPoint(pm.databaseContext) {
+			genesis, err := pm.dagTraversalManager.HighestChainBlockBelowBlueScore(model.VirtualBlockHash, pm.pruningDepth+pm.finalityInterval)
+			if err != nil {
+				return nil, err
+			}
+			utxoGenesisIter, err := pm.consensusStateManager.RestorePastUTXOSetIterator(genesis)
+			if err != nil {
+				return nil, err
+			}
+			serializedUtxo, err := serializeUTXOSetIterator(utxoGenesisIter)
+			if err != nil {
+				return nil, err
+			}
+			pm.pruningStore.Stage(genesis, serializedUtxo)
+			pruningPoint = genesis
+		} else {
+			return nil, err
+		}
+
+	}
+	return pruningPoint, nil
 }
 
 // SerializedUTXOSet returns the serialized UTXO set of the
 // current pruning point
 func (pm *pruningManager) SerializedUTXOSet() ([]byte, error) {
-	return nil, nil
+	return pm.pruningStore.PruningPointSerializedUTXOSet(pm.databaseContext)
+}
+
+func (pm *pruningManager) deletePastBlocks(pruningPoint *externalapi.DomainHash) error {
+	// Collect every node in highNode's past (including itself) but
+	// NOT in the lowNode's past (excluding itself) into an up-heap
+	// (a heap sorted by blueScore from lowest to greatest).
+	visited := map[externalapi.DomainHash]struct{}{}
+	queue := pm.dagTraversalManager.NewDownHeap()
+	parents, err := pm.dagTopologyManager.Parents(pruningPoint)
+	if err != nil {
+		return err
+	}
+	for _, parent := range parents {
+		err = queue.Push(parent)
+		if err != nil {
+			return err
+		}
+	}
+
+	for queue.Len() > 0 {
+		current := queue.Pop()
+		if _, ok := visited[*current]; ok {
+			continue
+		}
+		visited[*current] = struct{}{}
+
+		alreadyPruned, err := pm.deleteBlock(current)
+		if err != nil {
+			return err
+		}
+		if !alreadyPruned {
+			parents, err := pm.dagTopologyManager.Parents(current)
+			if err != nil {
+				return err
+			}
+			for _, parent := range parents {
+				err = queue.Push(parent)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (pm *pruningManager) deleteBlock(blockHash *externalapi.DomainHash) (alreadyPruned bool, err error) {
+	status, err := pm.blockStatusStore.Get(pm.databaseContext, blockHash)
+	if err != nil {
+		return false, err
+	}
+	if status == externalapi.StatusHeaderOnly {
+		return true, nil
+	}
+
+	pm.multiSetStore.Delete(blockHash)
+	pm.acceptanceDataStore.Delete(blockHash)
+	pm.blocksStore.Delete(blockHash)
+	pm.utxoDiffStore.Delete(blockHash)
+
+	pm.blockStatusStore.Stage(blockHash, externalapi.StatusHeaderOnly)
+	return false, nil
+}
+
+func pruningDepth(k, finalityInterval, mergeSetSizeLimit uint64) uint64 {
```

Comment:
Please move this function into dagconfig/params.go (next to FinalityDepth()) and pass it into PruningManager the same way you pass K.